### PR TITLE
Support multiple languages, in particular D

### LINKED ISSUE #7: Generalize to demangle other languages
It would be great if this mode was generalized to cover demangling of other languages, in my case, Dlang.

Such a generalization would mainly consist of stripping the sub-string `c++-filt` from symbols

-  `demangle-c++filt-executable`, and
- `demangle-c++filt-args`, and
- demangle--c++filt-command, 

and making these mode-local and maybe also buffer-local.

In D's case this executable is named `"ddemangle"` ([source](https://github.com/dlang/tools/blob/master/ddemangle.d)) instead of `"c++filt"`.

### PR DESCRIPTION
**Pull request by liblit**
[Mangled D symbols begin with `_D`.](https://dlang.org/spec/abi.html#name_mangling)

However, D can also refer to external symbols with C++ linkage, and therefore with C++ mangling. Thus we really need to support mangled symbols from multiple languages intermixed within a single buffer. We do this by maintaining distinct regexp patterns for each language, and associating each language with a filter command to use when demangling for that language. Thus, instead of a single demangler transaction queue we may have many:  one for each distinct filter command.

Fixes #7.

### CODE REVIEW

**Comment by liblit** [comment]
@nordlow, can I ask you to take this version of `demangle-mode` for a spin and tell me whether it does what you need for mangled D symbols? Once you confirm that it works well for you, I'll merge it into the master branch and cut a new release.

File: demangle-mode.el [hunk 1]
```diff
@@ -53,28 +53,34 @@
...
 
-(defcustom demangle-c++filt-executable "c++filt"
-  "The executable to use to decipher mangled symbols."
+(defcustom demangle-languages
+  '(("C++"
+     "\\(?:_Z\\|\\(?:_GLOBAL_[._$][DI]_\\)\\)[_$[:alnum:]]+"
+     ("c++filt" "--no-strip-underscore"))
+    ("D"
+     "_D[_[:alnum:]]+"
+     ("c++filt" "--no-strip-underscore" "--format=dlang")))
```
Comment (nordlow) [inline demangle-mode.el:73]: This should be configurable to use `ddemangle` as an alternative.

Comment (nordlow) [reply demangle-mode.el:73]: Update: We can always customize `demangle-languages` to use `ddemangle` so I guess this is a good default. Thanks!

**Comment by nordlow** [comment]
Thanks!

**Comment by liblit** [comment]
You're welcome, @nordlow! I hope that you and the rest of the D developer community find these changes useful.

**Comment by nordlow** [comment]
I announced it https://forum.dlang.org/post/xsopjcikfvrianssqosj@forum.dlang.org.

**Comment by nordlow** [comment]
I just realized there's a need for having both C++ and D demangling in stacktraces in D applications making use of `extern(C++) class`. For instances stacktraces in failing asserts inside of `dmd` needs this.

**Comment by BorisCarvajal** [comment]
Thanks for the feature!

> Can we arrange to have multiple demanglers activate in the same buffer?

@nordlow I've just tested this package following your announcement and the two languages are demangled just fine in the same buffer.

**Comment by liblit** [comment]
@nordlow wrote:
> Can we arrange to have multiple demanglers activate in the same buffer?

That already works. I did it that way from the start for exactly the reason you describe. When I saw that D uses C++ mangling for `extern(C++)` symbols, I could tell it would be good to let all demangling schemes be active simultaneously.

@BorisCarvajal wrote:
> Thanks for the feature!

You're welcome!

**Pull request merged**